### PR TITLE
Visual Studio 2015 RC support

### PR DESCRIPTION
**Pull request by tonyabbott**
This is my initial attempt at adding support for Visual Studio 2015 RC. I'm new to Gradle development and not an expert on Visual Studio so comments and suggestions are all welcome.

Through a bit of experimentation and some help from this blog post (http://blogs.msdn.com/b/vcblog/archive/2015/03/03/introducing-the-universal-crt.aspX) it seems the major difference that breaks Gradle support is the inclusion of the Universal CRT, which requires extra include and library paths. I haven't been able to find much other information on the UCRT, everything just seems to point back to that blog post.

The DefaultVisualStudioLocator and DefaultWindowsSdkLocator both work fine as-is. Windows Kit 8.1 must still be included as before, so no changes were required there. It seems to me to make sense to introduce the concept of the UCRT into Gradle, which is the approach I have taken. The DefaultUcrtLocator will look either in the registry or a user-specified location, there isn't really an option of locating it through the system path as there are no executables in there. I have made locating a UCRT optional, if it does not find one it will just not add any paths. This allows it to be backwards compatible with previous VS versions, but could potentially result in confusing missing headers or linker errors if it for some reason didn't locate the UCRT properly.

The UCRT is by default installed under `C:\Program Files (x86)\Windows Kits\10`, but the layout under that is slightly different from the other Windows Kits directories in that there is also a version directory under include. The default installation location from VS2015 RC will result in adding `C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt` to the include path and `C:\Program Files (x86)\Windows Kits\10\Lib\10.0.10150.0\ucrt\x64` to the library path. The version number 10.0.10150.0 changed between CT6 and the RC, so I have assumed it can change and handled that accordingly in the locator. I don't know if there is a possibility of having multiple versions of the UCRT installed concurrently, but have catered for that in the code too, picking the greatest version found as the default. We could perhaps allow the user to override that, but as I'm not even sure if it's a valid use case I haven't included that ability for now.

With regards platform support, from my installation it appears that:
- Windows Kit 8.1 has support for arm, x64 and x86 but not arm64
- The UCRT (aka Windows Kit 10?) has support for arm, arm64, x64 and x86
As there is not yet support for arm64 in Gradle, I have disregarded it for now. I'm not sure how that would work anyway, with no support for it in Windows Kit 8.1 - perhaps someone else knows more?
ia64 is supported by Gradle, but as there is no support for it in VS2015 so I have disregarded it. I think the existing behaviour of the DefaultWindowsSdkLocator will just end up adding a non-existent library directory and so result in a linker error. I haven't handled the possibility explicitly in locating the UCRT, should this raise an exception or is it okay to just ignore the possibility and let the user see the resultant linker error?

Using this patch I've been able to compile all the native samples as well as some other projects. I have also added cunit and google-test libraries compiled using VS2015 RC so that those samples can be used with VS2015.

Comments, suggestions?

### CODE REVIEW

**Comment by sherter** [comment]
Fixes [#GRADLE-3313](https://issues.gradle.org/browse/GRADLE-3313), much appreciated!

**Comment by Diom** [comment]
Yeah, I've also used it to work with VS2015 and it works a treat.

**Comment by big-guy** [comment]
FYI, we're working on our Windows test infrastructure to support testing VS2013/2015. Once we have that done, we'll be able to process this PR.

File: subprojects/platform-native/src/main/groovy/org/gradle/nativeplatform/toolchain/internal/msvcpp/DefaultUcrtLocator.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            libDirs.put(dir.getName(), dir);
+        }
+        Set<String> ucrtVersions = new HashSet<String>();
+        ucrtVersions.addAll(includeDirs.keySet());
+        ucrtVersions.addAll(libDirs.keySet());
+
+        List<String> result = new ArrayList<String>();
+        for (String version : ucrtVersions) {
+            File inc = includeDirs.get(version);
+            File lib = includeDirs.get(version);
```
Comment (mlipper) [inline subprojects/platform-native/src/main/groovy/org/gradle/nativeplatform/toolchain/internal/msvcpp/DefaultUcrtLocator.java]: Are both the 'inc' (line no. 158) and 'lib' (line no. 159) variables meant to reference the same Map.Entry<String, File> instance in the 'includeDirs' variable (line no. 144)? I'm guessing that the following was intended:

```java
File inc = includeDirs.get(version);
File lib = libDirs.get(version);
```

Comment (mlipper) [inline subprojects/platform-native/src/main/groovy/org/gradle/nativeplatform/toolchain/internal/msvcpp/DefaultUcrtLocator.java]: Are both the 'inc' (line no. 158) and 'lib' (line no. 159) variables meant to reference the same Map.Entry<String, File> instance in the 'includeDirs' variable (line no. 144)? I'm guessing that the following was intended:

```java
File inc = includeDirs.get(version);
File lib = libDirs.get(version);
```

**Comment by pioterj** [comment]
BTW the above-mentioned work on test infrastructure is in progress.

**Comment by pioterj** [comment]
The work on the performance test infrastructure is completed. I'm marking it as waiting for Gradle team's response as the engineer that owns this PR is currently not available.

**Comment by chefhoobajoob** [comment]
Has anybody tested Visual Studio solution and project file generation tasks with this fix?

**Comment by unjello** [comment]
hey guys, any forecasts on when this could get pulled in officially?

**Comment by ghale** [comment]
This has not fallen off the radar, and we're keen to add the support.  There is a little work that needs to be done before we can merge this and we expect to start that soon.  I can't provide a concrete date, but it shouldn't be much longer.  Thanks for the continued patience!

**Comment by hcoona** [comment]
Looking forward to this MR. Visual C++ in 2015 support latest C++11 which is quite helpful in development.

**Comment by zosrothko** [comment]
hey guys, any forecasts on when this could get pulled in officially? This PR is one year old ;)

**Comment by jjohannes** [comment]
Superseded by https://github.com/gradle/gradle/pull/695

**Pull request closed without merging**